File: prism-kotlin-sdk/credentials/src/commonMain/kotlin/io/iohk/atala/prism/kotlin/credentials/utils/Mustache.kt
package io.iohk.atala.prism.kotlin.credentials.utils

import com.github.h0tk3y.betterParse.combinators.*
import com.github.h0tk3y.betterParse.grammar.*
import com.github.h0tk3y.betterParse.lexer.*
import com.github.h0tk3y.betterParse.parser.*

sealed class MustacheNode
data class Literal(val content: String) : MustacheNode()
data class Comment(val content: String) : MustacheNode()
data class Variable(val name: String, val escape: Boolean) : MustacheNode()

data class Template(val tags: List<MustacheNode>) {
    /**
     * Validate template with the given context.
     */
    fun validate(context: TemplateContext): List<MustacheValidationError> {
        return tags.mapNotNull { tag ->
            when (tag) {
                is Variable ->
                    if (context(tag.name) == null) MustacheValidationError("Variable not found: ${tag.name}") else null

                else -> null
            }
        }
    }
}

sealed class MustacheError(message: String) : Exception(message)
class MustacheParsingError(message: String) : MustacheError(message)
data class MustacheValidationError(
    override val message: String,
    val errors: List<MustacheValidationError> = emptyList()
) : MustacheError(message)

typealias TemplateContext = (String) -> String?

/**
 * Subset of Mustache templates (https://mustache.github.io/) for internal usage in Prism SDK.
 *
 * Implemented specification follows https://mustache.github.io/mustache.5.html
 *
 * Currently supported tags:
 *   - variables: {{ variable }} or with path {{ variable.name }}
 *   - unescaped variables: {{& variable }}
 *   - comments with multiline support: {{! variable }}
 *
 * Usage:
 * {{{
 *   val context = { variable: String -> "$variable value" }
 *   Mustache.render("Template with {{ variable }}.", context) == "Template with variable value."
 * }}}
 */
object Mustache {
    val mustacheGrammar = object : Grammar<Template>() {
        val open by literalToken("{{")
        val close by literalToken("}}")
        val dot by literalToken(".")
        val andSign by literalToken("&")
        val bangSign by literalToken("!")
        val ws by regexToken("\\s+")
        val identifierToken by regexToken("([a-zA-Z\\d_]*)")
        val anyCharacter by regexToken(".|\r|\n")

        val literalString by oneOrMore(
            ws or dot or andSign or bangSign or identifierToken or anyCharacter
        ) use { joinToString("") { it.text } }
        val literal by literalString map { Literal(it) }

        val identifier by separatedTerms(
            identifierToken,
            dot,
            acceptZero = true
        ) use { joinToString(".") { it.text } }

        val variable by mustache(pad(identifier) map { Variable(it, escape = true) })

        val unescapedVariable by mustache(-andSign and pad(identifier) map { Variable(it, escape = false) })

        val comment by mustache(-bangSign and pad(literalString) map { Comment(it.trim()) })

        val statements by variable or unescapedVariable or comment

        override val rootParser: Parser<Template>
            get() = zeroOrMore(literal or statements) map { Template(it) }

        private fun pad(parser: Parser<String>): Parser<String> =
            -optional(ws) and parser and -optional(ws)

        private fun mustache(parser: Parser<MustacheNode>): Parser<MustacheNode> =
            -open and parser and -close
    }

    fun render(template: Template, context: TemplateContext, validate: Boolean): String {
        if (validate) {
            val errors = template.validate(context)
            if (errors.isNotEmpty()) {
                throw MustacheValidationError(
                    "Given context doesn't contain all variables from the template.",
                    errors
                )
            }
        }

        return template.tags.joinToString("") {
            when (it) {
                is Literal -> it.content
                is Comment -> ""
                is Variable ->
                    when (it.escape) {
                        true -> escapeHtml(context(it.name) ?: "")
                        false -> context(it.name) ?: ""
                    }
            }
        }
    }

    fun render(template: String, context: TemplateContext, validate: Boolean = true): String =
        try {
            render(mustacheGrammar.parseToEnd(template), context, validate)
        } catch (e: ParseException) {
            throw MustacheParsingError(e.message!!)
        }

    /**
     * Escape html entities.
     * Solution from: https://github.com/janl/mustache.js/blob/master/mustache.js#L67-L76
     */
    fun escapeHtml(html: String): String =
        buildString {
            html.forEach {
                when (it) {
                    '\'' -> append("&#39;")
                    '\"' -> append("&quot;")
                    '&' -> append("&amp;")
                    '<' -> append("&lt;")
                    '>' -> append("&gt;")
                    '/' -> append("&#x2F;")
                    '`' -> append("&#x60;")
                    '=' -> append("&#x3D;")
                    else -> append(it)
                }
            }
        }
}

File: prism-sdk/credentials/shared/src/main/scala/io/iohk/atala/prism/credentials/utils/Mustache.scala
package io.iohk.atala.prism.credentials.utils

import scala.util.parsing.input.Positional
import scala.util.parsing.combinator.JavaTokenParsers

import cats.data.ValidatedNel
import cats.implicits._

import io.iohk.atala.prism.credentials.utils.Mustache._

/**
  * Subset of Mustache templates (https://mustache.github.io/) for internal usage in Prism SDK.
  *
  * Implemented specification follows https://mustache.github.io/mustache.5.html
  *
  * Currently supported tags:
  *   - variables: {{ variable }} or with path {{ variable.name }}
  *   - unescaped variables: {{& variable }}
  *   - comments with multiline support: {{! variable }}
  *
  * Usage:
  *
  * {{{
  *   val mustache = new Mustache
  *   val context: TemplateContext = (variable: String) => Some(variable)
  *   mustache.render("Template with {{ variable }}.", context) == Right("Template with variable".)
  * }}}
  */
class Mustache extends JavaTokenParsers {

  override def skipWhitespace = false

  val open: Parser[String] = "{{"
  val close: Parser[String] = "}}"

  val literal: Parser[Literal] =
    positioned(rep1(not(open | close) ~> ".|\r|\n".r) ^^ { l => Literal(l.mkString) })

  val identifier: Parser[String] =
    rep1sep(ident, ".") ^^ { _.mkString(".") }

  val variable: Parser[Variable] =
    mustache(pad(identifier) ^^ { Variable(_, escape = true) })

  val unescapedVariable: Parser[Variable] =
    mustache("&" ~> pad(identifier) ^^ { Variable(_, escape = false) })

  val comment: Parser[Comment] =
    mustache("!" ~> pad(literal) ^^ { l => Comment(l.content.trim) })

  val statements: Parser[MustacheNode] =
    variable | unescapedVariable | comment

  val template: Parser[Template] =
    phrase(rep(literal | statements)) ^^ { Template(_) }

  /**
    * Parse string into template tree.
    */
  def parse(content: String): Either[MustacheError, Template] = {
    parse(template, content) match {
      case Success(matched, _) => Right(matched)
      case Failure(msg, _) => Left(MustacheParsingError(s"Template parsing failure: $msg"))
      case Error(msg, _) => Left(MustacheParsingError(s"Template parsing error: $msg"))
    }
  }

  /**
    * Render given template with provided context.
    *
    * @param validate Check if the given context contains all variables from the template.
    */
  def render(template: Template, context: TemplateContext, validate: Boolean): Either[MustacheError, String] = {
    for {
      _ <-
        if (validate) template.isValid(context).toEither.leftMap(_.head) // Return first validation error only.
        else Right(template)
    } yield template.tags.map {
      case Literal(content) => content
      case _: Comment => ""
      case Variable(variable, escape) =>
        context(variable) match {
          case Some(value) if escape => Mustache.escapeHtml(value.toString)
          case Some(value) if escape == false => value.toString
          case None => "" // if validate == false insert an empty string in place of nonexistent variables
        }
    }.mkString
  }

  /**
    * Render given string template with provided context.
    *
    * @param validate Check if the given context contains all variables from the template.
    */
  def render(content: String, context: TemplateContext, validate: Boolean = true): Either[MustacheError, String] =
    parse(content).flatMap(render(_, context, validate))

  /**
    * Render given string template with provided context.
    *
    * @throws MustacheError
    *
    * @param validate Check if the given context contains all variables from the template.
    */
  def unsafeRender(content: String, context: TemplateContext, validate: Boolean = true): String =
    render(content, context, validate).valueOr(throw _)

  /**
    * Trim whitespaces.
    */
  private def pad[T](parser: Parser[T]): Parser[T] =
    opt(whiteSpace) ~> parser <~ opt(whiteSpace)

  /**
    * Combine parser into mustache tags.
    */
  private def mustache[T <: MustacheNode](parser: Parser[T]): Parser[T] =
    positioned(open ~> parser <~ close)
}

object Mustache {
  sealed trait MustacheNode extends Positional
  case class Literal(content: String) extends MustacheNode
  case class Comment(content: String) extends MustacheNode
  case class Variable(name: String, escape: Boolean) extends MustacheNode

  case class Template(tags: Seq[MustacheNode]) {

    /**
      * Validate template with the given context, and return all errors or self.
      */
    def isValid(context: TemplateContext): MustacheProcessingResult[Unit] = {
      tags.toList
        .traverse {
          case variable @ Variable(name, _) =>
            Either
              .cond(context(name).isDefined, variable, MustacheValidationError(s"Variable not found: $name"))
              .toValidatedNel
          case other => other.validNel
        }
        .map(_ => ())
    }
  }

  sealed trait MustacheError extends Exception
  case class MustacheParsingError(message: String) extends MustacheError
  case class MustacheValidationError(message: String) extends MustacheError

  type MustacheProcessingResult[A] = ValidatedNel[MustacheError, A]

  type TemplateContext = String => Option[Any]

  /**
    * Render given string template with provided context.
    *
    * @param validate Check if the given context contains all variables from the template.
    */
  def render(content: String, context: TemplateContext, validate: Boolean = true): Either[MustacheError, String] =
    (new Mustache).render(content, context, validate)

  /**
    * Determine if given string is a Mustache template.
    */
  def isMustacheTemplate(content: String): Boolean =
    "\\{\\{(?:(?!}}).)*\\}\\}".r.findFirstIn(content).isDefined

  /**
    * Escape html entities.
    * Solution from: https://github.com/janl/mustache.js/blob/master/mustache.js#L67-L76
    */
  def escapeHtml(html: String): String = {
    html.map {
      case '\'' => "&#39;"
      case '\"' => "&quot;"
      case '&' => "&amp;"
      case '<' => "&lt;"
      case '>' => "&gt;"
      case '/' => "&#x2F;"
      case '`' => "&#x60;"
      case '=' => "&#x3D;"
      case other => other
    }.mkString
  }

}
